Lib/contextlib.py

```diff
@@ -471,6 +472,105 @@ def _fix_exception_context(new_exc, old_exc):
+            # the with statement, even if attempts to open files later
+            # in the list raise an exception
+    """
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *exc_details):
+        gen = self._shutdown_loop(*exc_details)
+        try:
+            result = next(gen)
+            while 1:
+                result = gen.send(result)
+        except StopIteration as e:
+            return e.value
+
+    def close(self):
+        """Immediately unwind the context stack"""
+        self.__exit__(None, None, None)
+
+
+# Inspired by discussions on https://bugs.python.org/issue29302
+class AsyncExitStack(_BaseExitStack):
+    """Async Context manager for dynamic management of a stack of exit callbacks
+    also maps __enter__ and __exit__ to outer __aenter__, __aexit__
+    For example:
+        with AsyncExitStack() as stack:
+                  connections = [await stack.enter_context(get_connection()) for i in range(5)]
+                  # All opened connections will automatically be released at the end of
+                  # the async with statement, even if attempts to open a connection later
+                  # in the list raise an exception
+    """
+
+    @staticmethod
+    def _create_exit_wrapper(cm, cm_exit):
+        if inspect.iscoroutinefunction(cm_exit):
+            async def _exit_wrapper(exc_type, exc, tb):
+                return await cm_exit(cm, exc_type, exc, tb)
+            return _exit_wrapper
+
+        return _BaseExitStack._create_exit_wrapper(cm, cm_exit)
+
+    @staticmethod
+    def _create_cb_wrapper(callback, *args, **kwds):
+        if inspect.iscoroutinefunction(callback):
+            async def _exit_wrapper(exc_type, exc, tb):
+                await callback(*args, **kwds)
+            return _exit_wrapper
+
+        return _BaseExitStack._create_cb_wrapper(callback, *args, **kwds)
+
+    async def enter_context(self, cm):
+        """Enters the supplied context manager
+        If successful, also pushes its __exit__ method as a callback and
+        returns the result of the __enter__ method.
+        """
+        # We look up the special methods on the type to match the with statement
+        _cm_type = type(cm)
+
+        # if you have both aenter + enter, aenter will 'win'
+        _exit = getattr(_cm_type, '__aexit__', None)
+        if _exit is None:
+            return super().enter_context(cm)
+
+        result = await _cm_type.__aenter__(cm)
+        self._push_cm_exit(cm, _exit)
+        return result
+
+    async def close(self):
+        """Immediately unwind the context stack"""
+        await self.__aexit__(None, None, None)
+
+    async def __aenter__(self):
+        return self
+
+    async def __aexit__(self, *exc_details):
+        gen = self._shutdown_loop(*exc_details)
+        try:
+            result = next(gen)
+            while 1:
+                try:
+                    if inspect.isawaitable(result):
```

Comment:
If we do decide to allow synchronous context managers in asynchronous exit stacks, I'd suggest wrapping their exit callbacks in an awaitable, so this check can be avoided.

Reply:
`inspect` is also used by `_create_exit_wrapper` and `_create_cb_wrapper`.